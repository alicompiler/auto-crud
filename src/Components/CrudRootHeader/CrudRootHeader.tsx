import React, {Component} from 'react';
import {CrudContextValue, CrudContext} from "../../Root/CrudContext";
import TitleComponent from "../TitleComponent/TitleComponent";
import {CrudRootHeaderDefault as Defaults} from "../../Defaults/Components/CrudRootHeaderDefautlts";
import {withRouter} from 'react-router-dom';


interface Props {
    history: any;
    match: any;
    location: any;
}

class CrudRootHeader extends Component<Props> {

    render() {
        return (
            <div>
                <div className={`__auto_crud_header ${Defaults.wrapperClassName}`}
                     style={{display: 'flex', justifyContent: 'space-between', padding: 6, alignItems: 'center'}}>
                    <TitleComponent/>
                    <div>
                        {this.renderHomeButton()}
                    </div>
                </div>
                <div className={'my-2 h-px w-full bg-gray-300'}/>
            </div>
        );
    }

    protected renderHomeButton = () => {
        if (Defaults.renderButton)
            return Defaults.renderButton();
        return <button className={`${Defaults.homeButtonClassName}`}
                       onClick={() => this.props.history.push(this.getContext().config.routeRoot)}>
            <i className={`${Defaults.homeButtonIconClassName}`}/>
        </button>
    };

    protected getContext = (): CrudContextValue => {
        return this.context;
    }
}

CrudRootHeader.contextType = CrudContext;

export default withRouter(CrudRootHeader);