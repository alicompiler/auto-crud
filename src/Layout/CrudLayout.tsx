import React, {Component} from 'react';
import {CrudContext, CrudContextValue} from "../Root/CrudContext";
import {Switch} from "react-router-dom";
import RoutesExtractor from "./RoutesExtractor";


class CrudLayout extends Component {

    private previousSwitch: any = null;

    render() {
        return (
            <div>
                {
                    this.getRouterSwitch()
                }
            </div>
        );
    }


    protected getRouterSwitch = () => {
        if (this.previousSwitch)
            return this.previousSwitch;
        const routesExtractor = new RoutesExtractor(this.getContext());
        this.previousSwitch = <Switch>
            {routesExtractor.getRoutes()}
        </Switch>;
        return this.previousSwitch;
    };

    protected getContext = (): CrudContextValue => this.context;

}

CrudLayout.contextType = CrudContext;

export default CrudLayout;