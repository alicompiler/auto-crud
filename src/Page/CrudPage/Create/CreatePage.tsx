import FormPage from "../../FormPage/FormPage";
import {SubmitConfig} from "raf-axios-submitter/dist/SubmitConfig"
import {AxiosResponse} from "axios";
import {AutoCrudDefaults} from "../../../Defaults/AutoCrudDefaults";

class CreatePage extends FormPage {

    getDefaultSubmitConfig(): Partial<SubmitConfig> {
        const onFail = this.getOptions().onFail;
        const onSuccess = this.getOptions().onSuccess;
        return {
            onFail: (e: any) => {
                let hookResult = undefined;
                onFail && (hookResult = onFail(this, e));
                if (hookResult === false) {
                    return;
                }
                this.updateStateForced({error: e});
            },
            onSuccess: (response: AxiosResponse) => {
                let hookResult = undefined;
                onSuccess && (hookResult = onSuccess(this, response));
                if (hookResult === false) {
                    return;
                }
                this.getFormRef()!.clear();
                this.updateStateForced({error: null});
            },
            changeLoadingStatus: true,
        }
    }

    public getDefaultHttpMethod = (): string => this.getOptions().httpMethod ?? AutoCrudDefaults.httpMethods.createRequest;

    public getDefaultPageTitle = () => AutoCrudDefaults.pageTitles.create;

}

export default CreatePage;