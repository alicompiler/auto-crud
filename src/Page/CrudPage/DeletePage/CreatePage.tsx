import FormPage from "../../FormPage/FormPage";

class CreatePage extends FormPage {
}

export default CreatePage;