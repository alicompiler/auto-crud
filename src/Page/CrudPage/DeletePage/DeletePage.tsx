import {AutoCrudDefaults} from "../../AutoCrudDefaults";
import {ActionPage, ActionPageLocalization} from "../../Base/ActionPage";

class DeletePage extends ActionPage {

    getDefaultPageTitle = () => AutoCrudDefaults.pageTitles.delete;

    public renderDefaultMessageComponent = () => {
        return AutoCrudDefaults.components.deleteMessage({
            disabled: this.getState().__loading,
            handleDelete: () => this.handleAction(),
            handleCancel: () => this.navigateToHome()
        });
    }

    public getDefaultHttpMethod = () => AutoCrudDefaults.httpMethods.deleteRequest;

    public getDefaultLocalization(): ActionPageLocalization {
        return {
            successMessage: AutoCrudDefaults.localization.delete_success_message,
            errorMessage: AutoCrudDefaults.localization.fail_to_delete_message
        }
    }

}

export default DeletePage;