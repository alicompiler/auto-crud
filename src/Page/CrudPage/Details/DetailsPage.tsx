import BaseCrudPage from "../../Base/BaseCrudPage/BaseCrudPage";
import {KeyValueComponent} from "react-keyvalue-ui";
import React from "react";
import {DetailsPageOptions} from "./DetailsPageOptions";
import {AutoCrudDefaults} from "../../../Defaults/AutoCrudDefaults";
import {PageActionRenderer} from "../../Base/ActionRenderer/PageActionRenderer";

class DetailsPage extends BaseCrudPage {
    public renderContent(): any {
        const item = this.getItem();
        if (!item) {
            return this.renderNoItem();
        }


        return <div>

            {this.renderActions()}

            <KeyValueComponent item={item}/>

        </div>;
    }

    private renderActions = () => {
        const actionRenderer = new PageActionRenderer(this.getContext(), this.getItem(), this.props.history.push, ["details"]);
        return <div>
            <div style={{display: 'flex', alignItems: 'center', justifyContent: 'flex-end'}}>
                {actionRenderer.render()}
            </div>
            <div className={AutoCrudDefaults.classNames.divider}/>
        </div>
    }

    private renderNoItem = () => {
        const customRender = this.getOptions().renderNoItem;
        if (customRender) {
            return customRender(this);
        }
        return AutoCrudDefaults.components.noItem({onAction: () => this.navigateToHome()});
    }

    public getItem = () => {
        return this.getState().__item;
    }

    public getOptions(): DetailsPageOptions {
        return super.getOptions();
    }
}

export default DetailsPage;