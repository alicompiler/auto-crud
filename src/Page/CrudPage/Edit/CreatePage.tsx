import FormPage from "../../FormPage/FormPage";

class EditPage extends FormPage {
}

export default EditPage;