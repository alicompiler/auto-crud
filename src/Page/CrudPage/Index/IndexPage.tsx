import {TablePage} from "../../CollectionPage/TablePage/TablePage";
import React from "react";
import ToolbarComponent from "../../../Components/PageComponents/ToolbarComponent";
import ToolbarNavigationAction from "../../../Defaults/Components/ToolbarNavigationAction";
import {IndexPageOptions} from "./IndexPageOptions";
import {AutoCrudDefaults} from "../../../Defaults/AutoCrudDefaults";

export class IndexPage extends TablePage {

    public renderContent(): any {
        return <div>
            {this.renderCollectionContainer()}
        </div>
    }

    getDefaultPageTitle = () => AutoCrudDefaults.pageTitles.index;

    protected renderDefaultToolbar = () => {
        return <ToolbarComponent onSearch={this.onSearch}
                                 noSearch={this.getOptions().noSearch}
                                 searchInputClassName={this.getOptions().searchInputClassName}
                                 wrapperClassName={this.getOptions().toolbarWrapperClassName}
                                 page={this}
                                 searchInputPlaceholder={this.getOptions().searchInputPlaceholder ?? AutoCrudDefaults.localization.search}
                                 actions={this.getToolbarAction()}/>
    }

    protected onSearch = (value: string) => {
        const onSearch = this.getOptions().onSearch;
        onSearch ? onSearch(value, this) : this.defaultOnSearch(value);
    };

    public defaultOnSearch(value: string): void {
        let endpointRoot = this.getContext().config.endpointRoot;
        const url = AutoCrudDefaults.endpoints.search(value, endpointRoot);
        this.updateDataSourceUrl(url);
    }

    public updateDataSourceUrl = (url: string) => {
        this.updateOptions({...this.getOptions(), dataSourceUrl: () => url}, () => {
            this.forceUpdate(() => this.getCollectionContainerRef()!.startDataFetch());
        });
    }

    public getToolbarAction = () => {
        const toolbarActions = this.getOptions().toolbarActions;
        if (toolbarActions) {
            return toolbarActions.map(action => () => action(this));
        }
        return [
            () => <ToolbarNavigationAction icon={AutoCrudDefaults.classNames.toolbarActions.createNavigationIcon}
                                           buttonClassName={AutoCrudDefaults.classNames.toolbarActions.createButtonClassName}
                                           navigateTo={this.pageConfigUtils.getPageConfigByName('create').route!}/>,

            () => <button className={AutoCrudDefaults.classNames.toolbarActions.refreshButtonClassName}
                          onClick={() => this.getCollectionContainerRef()?.startDataFetch()}>
                <i className={AutoCrudDefaults.classNames.toolbarActions.refreshButtonIcon}/>
            </button>
        ];
    };

    public getOptions(): IndexPageOptions {
        return super.getOptions();
    }

}