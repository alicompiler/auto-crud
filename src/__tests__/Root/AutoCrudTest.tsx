import AutoCrud from "../../Root/AutoCrud";
import * as React from "react";
import {configure, mount} from "enzyme";
import Adapter from "enzyme-adapter-react-16"
import {IndexPage} from "../../Page/CrudPage/Index/IndexPage";
import CreatePage from "../../Page/CrudPage/Create/CreatePage";
import UpdatePage from "../../Page/CrudPage/Update/UpdatePage";
import DeletePage from "../../Page/CrudPage/DeletePage/DeletePage";
import DetailsPage from "../../Page/CrudPage/Details/DetailsPage";
import {MemoryRouter} from "react-router-dom";
import CrudRootHeader from "../../Components/CrudRootHeader/CrudRootHeader";
import CrudLayout from "../../Root/CrudLayout";


configure({adapter: new Adapter()});

describe('AutoCrud', () => {

    it('should setup config', function () {
        let instance: AutoCrud | null = null;
        mount(
            <MemoryRouter>
                <AutoCrud ref={ref => instance = ref} endpointRoot={'http://localhost:8080'} name={'books'}
                          fields={[]}/>
            </MemoryRouter>
        );
        const context = instance!.getContextValue();
        expect(String(context.config)).toEqual({
            name: 'books',
            header: () => <CrudRootHeader/>,
            layout: () => <CrudLayout/>,
            endpointRoot: 'http://localhost:8080',
            routeRoot: '/books',
            fields: [],
            indexPage: {route: '/books', pageComponent: IndexPage, skip: false, options: {}, name: 'index'},
            createPage: {route: '/books/create', pageComponent: CreatePage, skip: false, options: {}, name: 'create'},
            updatePage: {route: '/books/edit', pageComponent: UpdatePage, skip: false, options: {}, name: 'update'},
            deletePage: {route: '/books/remove', pageComponent: DeletePage, skip: false, options: {}, name: 'delete'},
            detailsPage: {
                route: '/books/details',
                pageComponent: DetailsPage,
                skip: false,
                options: {},
                name: 'details'
            },
            pages: [],
        }.toString());
    });

    it('should setup ui state', function () {
        let instance: AutoCrud | null = null;
        mount(
            <MemoryRouter>
                <AutoCrud ref={ref => instance = ref} endpointRoot={'http://localhost:8080'} name={'books'}
                          fields={[]}/>
            </MemoryRouter>
        );
        const state = instance!.getContextValue().getState();
        expect(state.uiState).toEqual({
            pages: {
                index: {}, create: {}, update: {}, delete: {}, details: {}
            },
            modals: {}
        });
    });

    it('should update state', function (done) {
        let instance: AutoCrud | null = null;
        mount(
            <MemoryRouter>
                <AutoCrud ref={ref => instance = ref} endpointRoot={'http://localhost:8080'} name={'books'}
                          fields={[]}/>
            </MemoryRouter>
        );

        instance!.getContextValue().updateState({uiState: {pages: {index: {test: true}}}}, (state: any) => {
            expect(state.uiState.pages.index).toEqual({test: true});
            done();
        });
    });

    it('should update page options', function (done) {
        let instance: AutoCrud | null = null;
        mount(
            <MemoryRouter>
                <AutoCrud ref={ref => instance = ref} endpointRoot={'http://localhost:8080'} name={'books'}
                          fields={[]}/>
            </MemoryRouter>
        );

        instance!.getContextValue().updatePageOptions('index', {test: true}, () => {
            expect(instance!.getContextValue().config.indexPage.options).toEqual({test: true});
            expect(instance!.getContextValue().state.config.indexPage.options).toEqual({test: true});
            done();
        });
    });

});