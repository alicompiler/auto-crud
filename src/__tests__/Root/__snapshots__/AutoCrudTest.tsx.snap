// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoCrud should match snapshot 1`] = `ReactWrapper {}`;
